internal/storage/v2/clickhouse/internal/traces.go

```diff
@@ -0,0 +1,213 @@
+
+				linksTraceId.Append(linksTraceIds)
+				linksSpanId.Append(linksSpanIds)
+				linksTraceState.Append(linksTraceStates)
+				linksAttributes.Append(linksAttrs)
+			}
+		}
+	}
+
+	return proto.Input{
+		{Name: "Timestamp", Data: timestamp},
+		{Name: "TraceId", Data: traceId},
+		{Name: "SpanId", Data: spanId},
+		{Name: "ParentSpanId", Data: parentSpanId},
+		{Name: "TraceState", Data: traceState},
+		{Name: "SpanName", Data: spanName},
+		{Name: "SpanKind", Data: spanKind},
+		{Name: "ServiceName", Data: serviceName},
+		{Name: "ResourceAttributes.keys", Data: resourceAttributesKeys},
+		{Name: "ResourceAttributes.values", Data: resourceAttributesValues},
+		{Name: "ScopeName", Data: scopeName},
+		{Name: "ScopeVersion", Data: scopeVersion},
+		{Name: "SpanAttributes.keys", Data: spanAttributesKeys},
+		{Name: "SpanAttributes.values", Data: spanAttributesValues},
+		{Name: "Duration", Data: duration},
+		{Name: "StatusCode", Data: statusCode},
+		{Name: "StatusMessage", Data: statusMessage},
+		{Name: "Events.Timestamp", Data: eventsTimestamp},
+		{Name: "Events.Name", Data: eventsName},
+		{Name: "Events.Attributes", Data: eventsAttributes},
+		{Name: "Links.TraceId", Data: linksTraceId},
+		{Name: "Links.SpanId", Data: linksSpanId},
+		{Name: "Links.TraceState", Data: linksTraceState},
+		{Name: "Links.Attributes", Data: linksAttributes},
+	}
+}
+
+func attributesToMap(attributes pcommon.Map) map[string]string {
+	result := map[string]string{}
+	attributes.Range(func(k string, v pcommon.Value) bool {
+		result[k] = v.AsString()
+		return true
+	})
+	return result
+}
+
+func getServiceName(resAttr pcommon.Map) string {
+	var serviceName string
+	if v, ok := resAttr.Get(conventions.AttributeServiceName); ok {
+		serviceName = v.AsString()
+	}
+
+	return serviceName
+}
+
+func convertEvents(events ptrace.SpanEventSlice) (times []time.Time, names []string, attrs []map[string]string) {
+	for i := 0; i < events.Len(); i++ {
+		event := events.At(i)
+		times = append(times, event.Timestamp().AsTime())
+		names = append(names, event.Name())
+		attrs = append(attrs, attributesToMap(event.Attributes()))
+	}
+	return times, names, attrs
+}
+
+func convertLinks(links ptrace.SpanLinkSlice) link {
+	traceIDs := make([]string, 0, links.Len())
+	spanIDs := make([]string, 0, links.Len())
+	states := make([]string, 0, links.Len())
+	attrs := make([]map[string]string, 0, links.Len())
+
+	for i := 0; i < links.Len(); i++ {
+		link := links.At(i)
+		traceIDs = append(traceIDs, traceIDToHexOrEmptyString(link.TraceID()))
+		spanIDs = append(spanIDs, spanIDToHexOrEmptyString(link.SpanID()))
+		states = append(states, link.TraceState().AsRaw())
+		attrs = append(attrs, attributesToMap(link.Attributes()))
+	}
+	return link{traceIDs, spanIDs, states, attrs}
+}
+
+type link struct {
+	traceIDs    []string
+	spanIDs     []string
+	traceStates []string
+	attrs       []map[string]string
+}
+
+func traceIDToHexOrEmptyString(id pcommon.TraceID) string {
```

Comment:
this should be an error, empty IDs are not valid in trace data

Reply:
in other words, your code should assume they are always valid (to avoid unnecessary error handling)